### PR TITLE
[SPARK-43893][PYTHON][CONNECT] Non-atomic data type support in Arrow-optimized Python UDF

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Support non-atomic data types in input and output of Arrow-optimized Python UDF.

Non-atomic data types refer to: ArrayType, MapType, and StructType.

### Why are the changes needed?
Parity with pickled Python UDFs.

### Does this PR introduce _any_ user-facing change?
Non-atomic data types are accepted as both input and output of Arrow-optimized Python UDF.

For example,
```py
>>> df = spark.range(1).selectExpr("struct(1, struct('John', 30, ('value', 10))) as nested_struct")
>>> df.select(udf(lambda x: str(x))("nested_struct")).first()
Row(<lambda>(nested_struct)="Row(col1=1, col2=Row(col1='John', col2=30, col3=Row(col1='value', col2=10)))")
```

### How was this patch tested?
Unit tests.

SPARK-40307

### CODE REVIEW

File: python/pyspark/sql/pandas/serializers.py [hunk 1]
```diff
@@ -298,26 +299,39 @@ class ArrowStreamPandasUDFSerializer(ArrowStreamPandasSerializer):
...
     """
 
-    def __init__(self, timezone, safecheck, assign_cols_by_name, df_for_struct=False):
+    def __init__(
+        self,
+        timezone,
+        safecheck,
+        assign_cols_by_name,
+        df_for_struct=False,
+        eval_type=PythonEvalType.SQL_SCALAR_PANDAS_UDF,
```
Comment (ueshin) [inline python/pyspark/sql/pandas/serializers.py:308]: It shouldn't take `eval_type` here as the same as `df_for_struct` that is calculated from the `eval_type` in `worker.py`.

Comment (xinrong-meng) [reply python/pyspark/sql/pandas/serializers.py]: Good point, adjusted.

Comment (xinrong-meng) [reply python/pyspark/sql/pandas/serializers.py]: Docstrings on how `df_for_struct` and `struct_in_pandas` is set according to eval_types, as well as user-facing documentation on struct type behaviors would be helpful. I may work on it as a follow-up.

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/udf.py [hunk 2]
```diff
@@ -129,18 +127,12 @@ def _create_py_udf(
...
     except TypeError:
         is_func_with_args = False
-    is_output_atomic_type = (
-        not isinstance(return_type, StructType)
-        and not isinstance(return_type, MapType)
-        and not isinstance(return_type, ArrayType)
-    )
     if is_arrow_enabled:
-        if is_output_atomic_type and is_func_with_args:
+        if is_func_with_args:
```
Comment (allisonwang-db) [inline python/pyspark/sql/udf.py:135]: Just curious, why can't we enable arrow optimization for UDFs that do not have input arguments?

Comment (xinrong-meng) [reply python/pyspark/sql/udf.py:135]: Arrow optimization only makes sense when UDFs have inputs because it's an improvement on input (de)serialization.

**Comment by xinrong-meng** [comment]
Merged to master, thanks all!

File: python/pyspark/sql/udf.py (hunk 2 above)
Comment (allisonwang-db) [reply python/pyspark/sql/udf.py:135]: But does it hurt if we enable Arrow for functions without input arguments? Then we can eliminate this restriction and enable Arrow for all kinds of Python UDFs.

Comment (xinrong-meng) [reply python/pyspark/sql/udf.py:135]: Good point. The warning will be removed when we enable Arrow by default.

**Pull request closed without merging**